### PR TITLE
nuclear cylinder storage

### PR DESCRIPTION
**Pull request by quardbreak**
## Description of changes

Adds nuclear cylinder storage which allow to storage nuclear cylinders.
Difference from bay version:
1. Default type has no cylinders, but I added special subtype.
2. Mappers can configure how much nuclear cylinders will spawn/can holded inside it.

## Why and what will this PR improve

Allow us to storage nuclear stuff, instead using closets and crates.

## Authorship

Baystation12

### CODE REVIEW

File: code/game/machinery/self_destruct_storage.dm [hunk 1]
```diff
@@ -0,0 +1,125 @@
+/obj/machinery/nuclear_cylinder_storage
+	name = "nuclear cylinder storage"
+	desc = "It's a secure, armored storage unit embeded into the floor for storing the nuclear cylinders."
+	icon = 'icons/obj/machines/self_destruct_storage.dmi'
+	icon_state = "base"
+	anchored = TRUE
+	density = FALSE
+	req_access = list(access_heads_vault)
```
Comment (afterthought2) [inline code/game/machinery/self_destruct_storage.dm]: This looks like a bay-ism; machinery generally shouldn't set access.

File: code/game/machinery/self_destruct_storage.dm [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+/// Special subtype for mapping.
+/obj/machinery/nuclear_cylinder_storage/prefilled
+	prefilled = TRUE
+
+/obj/machinery/nuclear_cylinder_storage/emag_act(remaining_charges, mob/user, emag_source)
+	to_chat(user, SPAN_NOTICE("The card fails to do anything. It seems this device has an advanced encryption system."))
+	return NO_EMAG_ACT
+
+/obj/machinery/nuclear_cylinder_storage/physical_attack_hand(mob/user)
+	if(is_operable(src) && locked && check_access(user))
```
Comment (afterthought2) [inline code/game/machinery/self_destruct_storage.dm]: `is_operable` is deprecated, check flags directly.

File: code/game/machinery/self_destruct_storage.dm [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+		locked = FALSE
+		user.visible_message(
+			"\The [user] unlocks \the [src].", 
+			"You unlock \the [src]."
+		)
+		update_icon()
+		add_fingerprint(user)
```
Comment (afterthought2) [inline code/game/machinery/self_destruct_storage.dm]: Never add fingerprints anywhere. It's all handled for you.

File: code/game/machinery/self_destruct_storage.dm [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		)
+		if(do_after(user, 8 SECONDS, src) && open && (length(cylinders) < max_cylinders) && user.unEquip(O, src))
+			user.visible_message(
+				"\The [user] places \the [O] into storage.", 
+				"You place \the [O] into storage."
+			)
+			cylinders.Add(O)
+			update_icon()
+
+		add_fingerprint(user)
```
Comment (afterthought2) [inline code/game/machinery/self_destruct_storage.dm]: Call `..()` after this stuff (and don't add fingerprints).

File: code/game/machinery/self_destruct_storage.dm [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+/obj/machinery/nuclear_cylinder_storage/handle_mouse_drop(atom/over, mob/user)
+	if(over == user && open && length(cylinders))
+		user.visible_message(
+			"\The [user] begins to extract \the [cylinders[1]].", 
+			"You begin to extract \the [cylinders[1]]."
+		)
+		if(do_after(user, 7 SECONDS, src) && open && length(cylinders))
```
Comment (afterthought2) [inline code/game/machinery/self_destruct_storage.dm]: Why is the time randomly slightly less than the time to add them?

**Review by afterthought2 — COMMENTED** [review]
Please make it constructable.

**Comment by MistakeNot4892** [comment]
Why is this a machine?

File: code/game/machinery/self_destruct_storage.dm [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+/obj/machinery/nuclear_cylinder_storage/emag_act(remaining_charges, mob/user, emag_source)
+	to_chat(user, SPAN_NOTICE("The card fails to do anything. It seems this device has an advanced encryption system."))
+	return NO_EMAG_ACT
+
+/obj/machinery/nuclear_cylinder_storage/components_are_accessible(path)
+	return !locked && open && ..()
+
+/obj/machinery/nuclear_cylinder_storage/cannot_transition_to(state_path, mob/user)
+	if(locked)
+		return SPAN_NOTICE("You must unlock \the [src] first!")
```
Comment (MistakeNot4892) [inline code/game/machinery/self_destruct_storage.dm]: `SPAN_WARNING` for all of these.

File: code/game/machinery/self_destruct_storage.dm [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+	for(var/i in 1 to max_cylinders)
+		cylinders += new /obj/item/nuclear_cylinder()
+
+/// Special subtype for mapping.
+/obj/machinery/nuclear_cylinder_storage/prefilled
+	prefilled = TRUE
+
+/obj/machinery/nuclear_cylinder_storage/emag_act(remaining_charges, mob/user, emag_source)
+	to_chat(user, SPAN_NOTICE("The card fails to do anything. It seems this device has an advanced encryption system."))
```
Comment (MistakeNot4892) [inline code/game/machinery/self_destruct_storage.dm]: `SPAN_WARNING` here.

File: code/game/machinery/self_destruct_storage.dm [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+				"You place \the [O] into storage."
+			)
+			cylinders.Add(O)
+			update_icon()
+
+	..()
+
+/obj/machinery/nuclear_cylinder_storage/handle_mouse_drop(atom/over, mob/user)
+	if(over == user && open && !panel_open && length(cylinders))
+		user.visible_message(
```
Comment (MistakeNot4892) [inline code/game/machinery/self_destruct_storage.dm:126]: Would be better to cast cylinders[1] to a var and check that it's still in the cylinder list before removing, imo.

File: nebula.dme [hunk 9]
```diff
@@ -3347,6 +3349,7 @@
 #include "maps\antag_spawn\wizard\wizard.dm"
 #include "maps\away\away_sites.dm"
 #include "maps\away_sites_testing\away_sites_testing_define.dm"
+#include "maps\example\example.dm"
```
Comment (MistakeNot4892) [inline nebula.dme]: Erroneous include.

**Review by MistakeNot4892 — DISMISSED** [review]

File: code/game/machinery/self_destruct_storage.dm [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+				"You [open ? "open" : "close"] \the [src]."
+			)
+			update_icon()
+			return TRUE
+	else
+		to_chat(user, SPAN_WARNING("\The [src] is currently in maintenance mode!"))
+
+/obj/machinery/nuclear_cylinder_storage/attackby(obj/item/O, mob/user)
+	if(component_attackby(O, user))
+		return TRUE
```
Comment (afterthought2) [inline code/game/machinery/self_destruct_storage.dm]: It doesn't make sense to do both this and `..()` at the end.

**Review by afterthought2 — COMMENTED** [review]
I recommend just checking flags instead of `operable`, but I guess I won't force you.

**Review by afterthought2 — APPROVED** [review]

**Pull request merged**